<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.43">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>2&nbsp; DIPLOTOPIA pipeline and program descriptions – Diplotopia pipeline documentation</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./usage.html" rel="next">
<link href="./requirements_installation.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-d4d76bf8491c20bad77d141916dc28e1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-a34d670291f06f286357e447776a572a.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./description.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">DIPLOTOPIA pipeline and program descriptions</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Diplotopia pipeline documentation</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./requirements_installation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Requirements and installation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./description.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">DIPLOTOPIA pipeline and program descriptions</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./usage.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Input and usage</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./saga_users.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Saga Users</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./configuration.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Configuration</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./output.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Output files</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./contribution.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Contribution and support</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Summary</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Changes</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./changelog_versions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Change log in different versions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./changelog_development.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Change log</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Scientific reasoning</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./filter_contigs_choices.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Choices and reason behind those choices for : FILTER_CONTIGS track</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./haplopurge_choices.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Choices and reason behind those choices for : HAPLOPURGE track</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./VARWRRUMM_choices.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Choices and reason behind those choices for : VARWRRUMM track</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#sec-overview" id="toc-sec-overview" class="nav-link active" data-scroll-target="#sec-overview"><span class="header-section-number">2.1</span> Overview</a></li>
  <li><a href="#track-descriptions" id="toc-track-descriptions" class="nav-link" data-scroll-target="#track-descriptions"><span class="header-section-number">2.2</span> Track descriptions</a>
  <ul class="collapse">
  <li><a href="#purely_raw-cleaning-raw-reads-from-eventual-contaminants-status-dev" id="toc-purely_raw-cleaning-raw-reads-from-eventual-contaminants-status-dev" class="nav-link" data-scroll-target="#purely_raw-cleaning-raw-reads-from-eventual-contaminants-status-dev"><span class="header-section-number">2.2.1</span> PURELY_RAW: Cleaning raw reads from eventual contaminants (STATUS: DEV)</a></li>
  <li><a href="#tryssembly-testing-different-assembly-methods-status-dev" id="toc-tryssembly-testing-different-assembly-methods-status-dev" class="nav-link" data-scroll-target="#tryssembly-testing-different-assembly-methods-status-dev"><span class="header-section-number">2.2.2</span> TRYSSEMBLY: Testing different assembly methods (STATUS: DEV)</a></li>
  <li><a href="#filter_contigs-status-functional" id="toc-filter_contigs-status-functional" class="nav-link" data-scroll-target="#filter_contigs-status-functional"><span class="header-section-number">2.2.3</span> FILTER_CONTIGS (STATUS: FUNCTIONAL)</a></li>
  <li><a href="#compass-comparing-assemblies-status-functional" id="toc-compass-comparing-assemblies-status-functional" class="nav-link" data-scroll-target="#compass-comparing-assemblies-status-functional"><span class="header-section-number">2.2.4</span> COMPASS: Comparing assemblies (STATUS: FUNCTIONAL)</a></li>
  <li><a href="#haplopurge-status-functional" id="toc-haplopurge-status-functional" class="nav-link" data-scroll-target="#haplopurge-status-functional"><span class="header-section-number">2.2.5</span> HAPLOPURGE (STATUS: FUNCTIONAL)</a></li>
  <li><a href="#variant-calling-tracks" id="toc-variant-calling-tracks" class="nav-link" data-scroll-target="#variant-calling-tracks"><span class="header-section-number">2.2.6</span> Variant calling tracks</a></li>
  </ul></li>
  <li><a href="#fungen" id="toc-fungen" class="nav-link" data-scroll-target="#fungen"><span class="header-section-number">2.3</span> #### FunGen</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references"><span class="header-section-number">2.4</span> References</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">DIPLOTOPIA pipeline and program descriptions</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="sec-overview" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="sec-overview"><span class="header-section-number">2.1</span> Overview</h2>
<!-- learning graphviz https://quarto.org/docs/authoring/diagrams.html

https://graphviz.org/
graphviz online editor might help a bit
 -->
<div class="cell" data-layout-align="default">
<div class="cell-output-display">
<div id="fig-PipelineOverview" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-PipelineOverview-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<div>
<svg width="672" height="480" viewbox="0.00 0.00 421.75 609.20" xmlns="http://www.w3.org/2000/svg" xlink="http://www.w3.org/1999/xlink" style="; max-width: none; max-height: none">
<g id="graph0" class="graph" transform="scale(1 1) rotate(0) translate(4 605.2)">
<title>Pipeline_Overview</title>
<polygon fill="white" stroke="transparent" points="-4,4 -4,-605.2 417.75,-605.2 417.75,4 -4,4"></polygon>
<text text-anchor="middle" x="206.88" y="-584.6" font-family="Times,serif" font-size="14.00">Diplotopia Pipeline, Tracks Overview</text>
<g id="clust1" class="cluster">
<title>cluster_6</title>
<polygon fill="ghostwhite" stroke="ghostwhite" points="54.75,-8 54.75,-84.8 250.75,-84.8 250.75,-8 54.75,-8"></polygon>
<text text-anchor="middle" x="152.75" y="-68.2" font-family="Times,serif" font-size="14.00" fill="gray">Variant Calling Tracks</text>
</g>
<g id="clust3" class="cluster">
<title>cluster_666</title>
<path fill="none" stroke="gray" d="M285.75,-162.4C285.75,-162.4 393.75,-162.4 393.75,-162.4 399.75,-162.4 405.75,-168.4 405.75,-174.4 405.75,-174.4 405.75,-540.4 405.75,-540.4 405.75,-546.4 399.75,-552.4 393.75,-552.4 393.75,-552.4 285.75,-552.4 285.75,-552.4 279.75,-552.4 273.75,-546.4 273.75,-540.4 273.75,-540.4 273.75,-174.4 273.75,-174.4 273.75,-168.4 279.75,-162.4 285.75,-162.4"></path>
<text text-anchor="middle" x="339.75" y="-535.8" font-family="Times,serif" font-size="14.00" fill="gray">Legend</text>
</g>
<!-- PurelyrawT -->
<g id="node1" class="node">
<title>PurelyrawT</title>
<polygon fill="gray" stroke="gray" points="237.37,-522.2 28.14,-522.2 28.14,-481 237.37,-481 237.37,-522.2"></polygon>
<text text-anchor="middle" x="132.75" y="-505.8" font-family="Times,serif" font-size="14.00" fill="#97ffff">PURELY_RAW</text>
<text text-anchor="middle" x="132.75" y="-489" font-family="Times,serif" font-size="14.00" fill="#97ffff">(Reads decontamination; optional)</text>
</g>
<!-- TryssemblyT -->
<g id="node2" class="node">
<title>TryssemblyT</title>
<polygon fill="deeppink" stroke="deeppink" points="244.18,-444.6 21.32,-444.6 21.32,-403.4 244.18,-403.4 244.18,-444.6"></polygon>
<text text-anchor="middle" x="132.75" y="-428.2" font-family="Times,serif" font-size="14.00">TRYSSEMBLY</text>
<text text-anchor="middle" x="132.75" y="-411.4" font-family="Times,serif" font-size="14.00">(trial of different kind of assemblies)</text>
</g>
<!-- PurelyrawT&#45;&gt;TryssemblyT -->
<g id="edge1" class="edge">
<title>PurelyrawT-&gt;TryssemblyT</title>
<path fill="none" stroke="grey" stroke-dasharray="5,2" d="M132.75,-480.74C132.75,-471.27 132.75,-459.86 132.75,-449.83"></path>
<polygon fill="grey" stroke="grey" points="134.5,-449.73 132.75,-444.73 131,-449.73 134.5,-449.73"></polygon>
</g>
<!-- FiltercontigsT -->
<g id="node3" class="node">
<title>FiltercontigsT</title>
<polygon fill="#76ee00" stroke="#76ee00" points="227.26,-367 0.25,-367 0.25,-325.8 227.26,-325.8 227.26,-367"></polygon>
<text text-anchor="middle" x="113.75" y="-350.6" font-family="Times,serif" font-size="14.00" fill="#97ffff">FILTER_CONTIGS</text>
<text text-anchor="middle" x="113.75" y="-333.8" font-family="Times,serif" font-size="14.00" fill="#97ffff">(assembly decontamination; optional)</text>
</g>
<!-- TryssemblyT&#45;&gt;FiltercontigsT -->
<g id="edge2" class="edge">
<title>TryssemblyT-&gt;FiltercontigsT</title>
<path fill="none" stroke="black" d="M127.76,-403.14C125.38,-393.67 122.51,-382.26 119.99,-372.23"></path>
<polygon fill="black" stroke="black" points="121.63,-371.55 118.71,-367.13 118.23,-372.41 121.63,-371.55"></polygon>
</g>
<!-- CompassT -->
<g id="node4" class="node">
<title>CompassT</title>
<polygon fill="#76ee00" stroke="#76ee00" points="246.71,-211.8 76.79,-211.8 76.79,-170.6 246.71,-170.6 246.71,-211.8"></polygon>
<text text-anchor="middle" x="161.75" y="-195.4" font-family="Times,serif" font-size="14.00">COMPASS</text>
<text text-anchor="middle" x="161.75" y="-178.6" font-family="Times,serif" font-size="14.00">(comparison of assemblies)</text>
</g>
<!-- TryssemblyT&#45;&gt;CompassT -->
<g id="edge4" class="edge">
<title>TryssemblyT-&gt;CompassT</title>
<path fill="none" stroke="grey" stroke-dasharray="5,2" d="M194.74,-403.28C211.27,-394.88 227.18,-383.2 236.75,-367.2 263.95,-321.74 260.54,-295.34 236.75,-248 229.94,-234.45 218.11,-223.36 205.84,-214.73"></path>
<polygon fill="grey" stroke="grey" points="206.78,-213.26 201.66,-211.9 204.82,-216.16 206.78,-213.26"></polygon>
</g>
<!-- FiltercontigsT&#45;&gt;CompassT -->
<g id="edge5" class="edge">
<title>FiltercontigsT-&gt;CompassT</title>
<path fill="none" stroke="grey" stroke-dasharray="5,2" d="M100.12,-325.55C88.02,-305.26 74.07,-273.24 86.75,-248 93.56,-234.45 105.39,-223.36 117.67,-214.73"></path>
<polygon fill="grey" stroke="grey" points="118.69,-216.16 121.84,-211.9 116.72,-213.26 118.69,-216.16"></polygon>
</g>
<!-- HaplopurgeT -->
<g id="node5" class="node">
<title>HaplopurgeT</title>
<polygon fill="lightgreen" stroke="lightgreen" points="227.61,-289.4 95.89,-289.4 95.89,-248.2 227.61,-248.2 227.61,-289.4"></polygon>
<text text-anchor="middle" x="161.75" y="-273" font-family="Times,serif" font-size="14.00">HAPLOPURGE</text>
<text text-anchor="middle" x="161.75" y="-256.2" font-family="Times,serif" font-size="14.00">(get haploid contigs)</text>
</g>
<!-- FiltercontigsT&#45;&gt;HaplopurgeT -->
<g id="edge3" class="edge">
<title>FiltercontigsT-&gt;HaplopurgeT</title>
<path fill="none" stroke="black" d="M126.36,-325.54C132.54,-315.81 140.02,-304.03 146.51,-293.81"></path>
<polygon fill="black" stroke="black" points="148.02,-294.69 149.22,-289.53 145.07,-292.81 148.02,-294.69"></polygon>
</g>
<!-- Haplossembly -->
<g id="node8" class="node">
<title>Haplossembly</title>
<polygon fill="none" stroke="black" stroke-dasharray="1,5" points="223.86,-134.2 99.65,-134.2 99.65,-93 223.86,-93 223.86,-134.2"></polygon>
<text text-anchor="middle" x="161.75" y="-117.8" font-family="Times,serif" font-size="14.00">Choosen assembly </text>
<text text-anchor="middle" x="161.75" y="-101" font-family="Times,serif" font-size="14.00"> (HaploSSembly)</text>
</g>
<!-- CompassT&#45;&gt;Haplossembly -->
<g id="edge7" class="edge">
<title>CompassT-&gt;Haplossembly</title>
<path fill="none" stroke="black" d="M161.75,-170.34C161.75,-160.87 161.75,-149.46 161.75,-139.43"></path>
<polygon fill="black" stroke="black" points="163.5,-139.33 161.75,-134.33 160,-139.33 163.5,-139.33"></polygon>
</g>
<!-- HaplopurgeT&#45;&gt;CompassT -->
<g id="edge6" class="edge">
<title>HaplopurgeT-&gt;CompassT</title>
<path fill="none" stroke="black" d="M161.75,-247.94C161.75,-238.47 161.75,-227.06 161.75,-217.03"></path>
<polygon fill="black" stroke="black" points="163.5,-216.93 161.75,-211.93 160,-216.93 163.5,-216.93"></polygon>
</g>
<!-- FunGen -->
<g id="node6" class="node">
<title>FunGen</title>
<polygon fill="pink" stroke="pink" points="242.36,-52 181.14,-52 181.14,-16 242.36,-16 242.36,-52"></polygon>
<text text-anchor="middle" x="211.75" y="-29.8" font-family="Times,serif" font-size="14.00">FunGen</text>
</g>
<!-- VARWRRUM -->
<g id="node7" class="node">
<title>VARWRRUM</title>
<polygon fill="#76ee00" stroke="#76ee00" points="162.76,-52 62.75,-52 62.75,-16 162.76,-16 162.76,-52"></polygon>
<text text-anchor="middle" x="112.75" y="-29.8" font-family="Times,serif" font-size="14.00">VARWRRUM</text>
</g>
<!-- Haplossembly&#45;&gt;FunGen -->
<g id="edge8" class="edge">
<title>Haplossembly-&gt;FunGen</title>
<path fill="none" stroke="black" d="M174.63,-92.62C181.78,-81.52 190.68,-67.7 197.97,-56.39"></path>
<polygon fill="black" stroke="black" points="199.55,-57.17 200.78,-52.02 196.6,-55.28 199.55,-57.17"></polygon>
</g>
<!-- Haplossembly&#45;&gt;VARWRRUM -->
<g id="edge9" class="edge">
<title>Haplossembly-&gt;VARWRRUM</title>
<path fill="none" stroke="black" d="M149.13,-92.62C142.12,-81.52 133.4,-67.7 126.26,-56.39"></path>
<polygon fill="black" stroke="black" points="127.65,-55.32 123.5,-52.02 124.69,-57.19 127.65,-55.32"></polygon>
</g>
<!-- planned -->
<g id="node9" class="node">
<title>planned</title>
<polygon fill="gray" stroke="gray" points="370.07,-519.6 309.43,-519.6 309.43,-483.6 370.07,-483.6 370.07,-519.6"></polygon>
<text text-anchor="middle" x="339.75" y="-497.4" font-family="Times,serif" font-size="14.00">planned</text>
</g>
<!-- building -->
<g id="node10" class="node">
<title>building</title>
<polygon fill="pink" stroke="pink" points="397.95,-442 281.55,-442 281.55,-406 397.95,-406 397.95,-442"></polygon>
<text text-anchor="middle" x="339.75" y="-419.8" font-family="Times,serif" font-size="14.00">currently building</text>
</g>
<!-- planned&#45;&gt;building -->
<!-- partial -->
<g id="node11" class="node">
<title>partial</title>
<polygon fill="deeppink" stroke="deeppink" points="381.79,-364.4 297.71,-364.4 297.71,-328.4 381.79,-328.4 381.79,-364.4"></polygon>
<text text-anchor="middle" x="339.75" y="-342.2" font-family="Times,serif" font-size="14.00">partial build</text>
</g>
<!-- building&#45;&gt;partial -->
<!-- functional -->
<g id="node12" class="node">
<title>functional</title>
<polygon fill="#76ee00" stroke="#76ee00" points="376.01,-286.8 303.49,-286.8 303.49,-250.8 376.01,-250.8 376.01,-286.8"></polygon>
<text text-anchor="middle" x="339.75" y="-264.6" font-family="Times,serif" font-size="14.00">functional</text>
</g>
<!-- partial&#45;&gt;functional -->
<!-- optional -->
<g id="node13" class="node">
<title>optional</title>
<polygon fill="ghostwhite" stroke="ghostwhite" points="379.89,-211.8 299.61,-211.8 299.61,-170.6 379.89,-170.6 379.89,-211.8"></polygon>
<text text-anchor="middle" x="339.75" y="-195.4" font-family="Times,serif" font-size="14.00" fill="#97ffff">optional</text>
<text text-anchor="middle" x="339.75" y="-178.6" font-family="Times,serif" font-size="14.00" fill="#97ffff">(font color)</text>
</g>
<!-- functional&#45;&gt;optional -->
</g>
</svg>
</div>
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-PipelineOverview-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2.1: This is a simple graphviz graph.
</figcaption>
</figure>
</div>
</div>
</div>
<!-- would nbe better if nodes were of same size
https://stackoverflow.com/questions/48888032/graphviz-make-all-nodes-the-same-size-as-the-largest

would be good if in legend boxes were closer to each other
-->
<!--
See also:

General overview of the pipeline
 ![DIPLOTOPIA overview](./files/pipeline_overview.drawio.svg)

-->
</section>
<section id="track-descriptions" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="track-descriptions"><span class="header-section-number">2.2</span> Track descriptions</h2>
<!-- # general description -->
<!-- #  short description each track -->
<hr>
<section id="purely_raw-cleaning-raw-reads-from-eventual-contaminants-status-dev" class="level3" data-number="2.2.1">
<h3 data-number="2.2.1" class="anchored" data-anchor-id="purely_raw-cleaning-raw-reads-from-eventual-contaminants-status-dev"><span class="header-section-number">2.2.1</span> PURELY_RAW: Cleaning raw reads from eventual contaminants (STATUS: DEV)</h3>
<p><strong>Pre-assembly. Detection and filtering out (optional) of contaminated reads, prior to assembly.</strong> <!-- Description

- [ ] raw reads nanopore ? -> Thomas track integration ?
- [ ] raw reads illumina ?`

#### Tools used in this track and references

References
Repositories
Manuals and other resources

--></p>
<hr>
</section>
<section id="tryssembly-testing-different-assembly-methods-status-dev" class="level3" data-number="2.2.2">
<h3 data-number="2.2.2" class="anchored" data-anchor-id="tryssembly-testing-different-assembly-methods-status-dev"><span class="header-section-number">2.2.2</span> TRYSSEMBLY: Testing different assembly methods (STATUS: DEV)</h3>
<p><strong>Trial of different assemblers to obtain a suitable assembly for your purposes.</strong></p>
<p>As per today, assemblies are a model of an organisms’ genome, not the truth! Testing different assemblies methods (or assemblers) might serve your purpose. Indeed, some assemblers will perform better with some organisms, while some better is other organisms. Moreover, depending on what is the ultimate goal for you to do an assembly (eg. syntheny comparison of organisms belonging to the same species, copy number variation, variant calling ….) then some methods might serve your ultimate goal better than others.</p>
<p>As it is difficult to predict what will work best (at least for me), then it might be usefull to try different methods rapidely. Here is a track that aims to give you some assembly options.</p>
<p>(Note that this track is expected to growth …. when the need arise).</p>
<p>Currently the methods we are using here are:</p>
<ul>
<li>NECAT : from NECAT assembler specialized for long reads (nanopore), <a href="https://www.nature.com/articles/s41467-020-20236-7">NECAT ARTICLE</a> and <a href="https://github.com/xiaochuanle/NECAT">NECAT TOOL repository</a></li>
<li>MASURCA : from MASURCA genome assembler, an hybrid assembler, <a href="https://academic.oup.com/bioinformatics/article/29/21/2669/195975">MASURCA FIRST ARTICLE</a>, <a href="https://genome.cshlp.org/content/27/5/787">MASURCA HYBRID ARTICLE</a> and <a href="https://github.com/alekseyzimin/masurca">MASURCA TOOL repository</a></li>
</ul>
<!-- Description

#### Tools used in this track and references

References
Repositories
Manuals and other resources

-->
<hr>
</section>
<section id="filter_contigs-status-functional" class="level3" data-number="2.2.3">
<h3 data-number="2.2.3" class="anchored" data-anchor-id="filter_contigs-status-functional"><span class="header-section-number">2.2.3</span> FILTER_CONTIGS (STATUS: FUNCTIONAL)</h3>
<p>Optional track. <strong>Filters out contigs/scaffolds from assemblies, when those are not detected as belonging to a “positive” filtering criteria (eg. organism of interest). Employed for removal of contaminated contigs from assemblies.</strong> Main output: decontaminated asembly sequence</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./files/FILTER_CONTIGS.drawio.svg" class="img-fluid figure-img"></p>
<figcaption>FILTER_CONTIGS track overview</figcaption>
</figure>
</div>
<p>This track uses <a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0163962">seqkit</a> software to split assemblies into the set of contigs for each sample, allowing high parallelization of the [blastn] taxonomical comparison of each contig. Blastn is used to detect a maximum of 5 matches and to provide the taxonomical identification of those matches.</p>
<p>The blastn taxonomical results are then collated and used to filter out likely contaminant contig using the R-quarto script <code>diplotopia/bin/contigs_taxo_overview_filter.qmd</code>.</p>
<p>A “positive filter” expression (rlang) is provided by the user. Any taxonomy that is not reflected within the positive filter will be considered as contaminant. We aimed to make the choice of this expression as flexible as possible. It is eg. possible to use an expression composed of several taxonomic ranks, ensuring that we can fine tune the filtering step to obtain a “decontaminated” assembly.</p>
<!-- TODO check how the expression is build -->
<p>See <a href="./Filtering_contigs">Filtering contigs explanations</a> for more details about the filtering process strategy.</p>
<p>See also [Installation##Obtain_necessary_external databases], as this track require database/files to be pre-downloaded.</p>
<p>References - <a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0163962">SEQKIT article</a> - <a href="https://pubmed.ncbi.nlm.nih.gov/20003500/">BLAST article</a> <!-- this is the tool, where is the science ? --> - <a href="https://cran.r-project.org/">R</a> - <a href="https://quarto.org/">Quarto</a></p>
<!-- TODO session info / packages used in R script -->
<p>Repositories - <a href="https://github.com/shenwei356/seqkit/releases">SEQKIT repository</a></p>
<p>Manuals and other resources - <a href="https://blast.ncbi.nlm.nih.gov/doc/blast-help/">NCBI Blast help</a></p>
<hr>
</section>
<section id="compass-comparing-assemblies-status-functional" class="level3" data-number="2.2.4">
<h3 data-number="2.2.4" class="anchored" data-anchor-id="compass-comparing-assemblies-status-functional"><span class="header-section-number">2.2.4</span> COMPASS: Comparing assemblies (STATUS: FUNCTIONAL)</h3>
<p><strong>Assembly comparison and evaluation - a guide to choice which assembly to use for further analyses</strong></p>
<p>This track aims to provide statistical information allowing to compare assemblies, in order to be able to make informed choice about which assembly you will use for the rest of your analyses, this depending of your purpose / criteria.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./files/COMPASS.drawio.svg" class="img-fluid figure-img"></p>
<figcaption>COMPASS track overview</figcaption>
</figure>
</div>
<p>The tracks provide statistical information (main output) about: - coverage (base coverage : deeph) (QUALIMAP) with results summarized with (MULITQC) - contiguity and coverage genome (relative to a given reference genome, which can be either one assembly you select or an external assembly, or both) - In some cases it can detect missassemblies, dependent of the reference genome used (QUAST) - completeness (BUSCO): it is possible to use one or several sets of markers. &gt; Eg. if no specific marker are developed for the lineage that is your object of interest, you can chose to use markers from several lineages for the comparison, and see which will provide the most informative completness informaiton.</p>
<!-- TODO : other things that should be added ? -->
<p>A summary table for completness (BUSCO) is generated using a R-quarto script (see diplotopia/bin/busco_summary.qmd) at the end of the track.</p>
<p>Note that it is also possible to run FILTER_CONTIGS track in COMPASS, this is for convenience during rerun, and when you are sure you have defined your filtering expression correctly.</p>
<!-- TODO put this part as optional ! in the pipeline-->
<p>Eventual contamination of the assemblies is checked using a kmer approach (KRAKEN2). Note that this information is purely indicative (and a deviation of normal KRAKEN2 usage). Moreover if the organism that you are interested in in NOT included in the kraken database that is used, this will most certainly report erroneous taxon content.</p>
<p>See also [Installation##Obtain_necessary_external databases], as this track require a kraken database to be either downloaded or created specifically for your analysis. <!-- TODO check that the infor is available and correct here --></p>
<p>References</p>
<ul>
<li><a href="https://academic.oup.com/bioinformatics/article/31/19/3210/211866">BUSCO article</a></li>
<li><a href="https://academic.oup.com/bioinformatics/article/29/8/1072/228832">QUAST article</a></li>
<li><a href="https://academic.oup.com/bioinformatics/article/28/20/2678/206551">QUALIMAP article</a></li>
<li><a href="https://academic.oup.com/bioinformatics/article/32/19/3047/2196507">MULTIQC article</a></li>
<li><a href="https://academic.oup.com/gigascience/article/10/2/giab008/6137722?login=false">SAMTOOLS article</a></li>
<li><a href="https://academic.oup.com/bioinformatics/article/34/18/3094/4994778">MINIMAP article</a></li>
<li><a href="https://arxiv.org/abs/1303.3997">BWA-MEM article</a></li>
<li><a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1891-0">KRAKEN2 article</a></li>
</ul>
<p>Repositories</p>
<ul>
<li><a href="https://github.com/ablab/quast">QUAST repository</a></li>
<li><a href="https://bitbucket.org/kokonech/qualimap/commits/branch/master">QUALIMAP repository</a></li>
<li><a href="https://gitlab.com/ezlab/busco">BUSCO repository</a></li>
<li><a href="https://github.com/MultiQC/MultiQC">MULTIQC repository</a></li>
<li><a href="https://github.com/samtools/samtools">SAMTOOLS repository</a></li>
<li><a href="https://github.com/lh3/minimap2">MINIMAP repository</a></li>
<li><a href="https://github.com/lh3/bwa">BWA repository</a></li>
<li><a href="https://github.com/DerrickWood/kraken2">KRAKEN2 repository</a></li>
</ul>
<!-- TODO need to add pavian ? make a pavian like output -->
<p>Manuals and other resources</p>
<ul>
<li><a href="https://busco.ezlab.org/busco_userguide.html#installation-with-docker">BUSCO manual</a></li>
</ul>
<hr>
</section>
<section id="haplopurge-status-functional" class="level3" data-number="2.2.5">
<h3 data-number="2.2.5" class="anchored" data-anchor-id="haplopurge-status-functional"><span class="header-section-number">2.2.5</span> HAPLOPURGE (STATUS: FUNCTIONAL)</h3>
<p><strong>From assembly to haploid assembly - retaining the haploid genome for variant calling</strong></p>
<p>! Currently only for long reads</p>
<!-- TODO  make it available for short reads (prioritized long reads because was was most likely used for obtaining good reference) -->
<p>The purpose of this track is to clean the chosen assembly to optain an haploid representation of the genome of a diploid organism. Ie, we aim to optain an haploid representation of the genome of a diploid organism. This haploid representation (HaploSSembly) will then be used to call variants (or any other aim where it fits) of diploid organisms, allow allowing to determine the genotype at each haploid locus. Main output: HaploSSembly (Haploid Assembly)</p>
<p>The original assembly, provided as input, is inspected for eventual chimera/missassembly contigs, junk (potential low copy number contaminants), repeats and haplotigs using <a href="" title="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-018-2485-7">purge haplotigs</a> software. <!-- TODO re-read paper: define better  --></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./files/HAPLOPURGE.drawio.svg" class="img-fluid figure-img"></p>
<figcaption>HAPLOPURGE track overview</figcaption>
</figure>
</div>
<p>First, reads are mapped (using MINIMAP) to the (decontaminated) assembly <!-- if you provide it as input - should be -->. Purge Haplotigs software allows to draw an histogram that allows to evaluate the coverage of the haploid and diploid parts of the genome, which allows to determine three coverage thresholds low, mid (between the haploid and diploid pics) and high that Purge Haploitigs software will further used to attemps to classify contigs as haploid VS diploid, poliploid (eg. small repeats) or chimeric/artificial contigs, this on the basis of the proportion of haploid and diploid coverage for each contig. <!-- TODO understand better and recheck info, and reformulate --></p>
<p>Based on this classification, haplotigs and other ambiguous contigs are discared from the original input assembly.</p>
<!-- TODO  : THINK
NB: it also allows, by examining the difference to the completeness and size of the assembly previously obtained, to evaluate how many alleles were assembled separately
- would be an estimate
- eg can have chimeaa / remove the complete version instead of incomplete version of a gene ?
? could it be used to provide a general idea of polymorphism at the level it was given before for sapro ? hein ... brrr
-->
<p>See also <a href="./Haplopurge_choices">HAPLOPURGE choices</a></p>
<p>References</p>
<ul>
<li><a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-018-2485-7">Purge haplotigs article</a></li>
<li><a href="https://academic.oup.com/bioinformatics/article/34/18/3094/4994778">MINIMAP article</a></li>
</ul>
<p>Repositories - <a href="https://bitbucket.org/mroachawri/purge_haplotigs/src/master/">Purge haplotigs repository</a> - <a href="https://github.com/lh3/minimap2">MINIMAP repository</a></p>
<p>Manuals and other resources</p>
<hr>
</section>
<section id="variant-calling-tracks" class="level3" data-number="2.2.6">
<h3 data-number="2.2.6" class="anchored" data-anchor-id="variant-calling-tracks"><span class="header-section-number">2.2.6</span> Variant calling tracks</h3>
<div class="cell" data-layout-align="default">
<div class="cell-output-display">
<div>
<p></p><figure class="figure"><p></p>
<div>
<svg width="672" height="480" viewbox="0.00 0.00 202.00 142.40" xmlns="http://www.w3.org/2000/svg" xlink="http://www.w3.org/1999/xlink" style="; max-width: none; max-height: none">
<g id="graph0" class="graph" transform="scale(1 1) rotate(0) translate(4 138.4)">
<title>G</title>
<polygon fill="white" stroke="transparent" points="-4,4 -4,-138.4 198,-138.4 198,4 -4,4"></polygon>
<text text-anchor="middle" x="97" y="-117.8" font-family="Times,serif" font-size="14.00">Variant Calling</text>
<text text-anchor="middle" x="97" y="-101" font-family="Times,serif" font-size="14.00">Tracks Overview</text>
<g id="clust1" class="cluster">
<title>cluster_0</title>
<polygon fill="pink" stroke="pink" points="8,-8 8,-84.8 78,-84.8 78,-8 8,-8"></polygon>
<text text-anchor="middle" x="43" y="-68.2" font-family="Times,serif" font-size="14.00">FungGen</text>
</g>
<g id="clust2" class="cluster">
<title>cluster_1</title>
<polygon fill="#76ee00" stroke="#76ee00" points="86,-8 86,-84.8 186,-84.8 186,-8 86,-8"></polygon>
<text text-anchor="middle" x="136" y="-68.2" font-family="Times,serif" font-size="14.00">VARWRRUM</text>
</g>
<!-- tata -->
<g id="node1" class="node">
<title>tata</title>
<polygon fill="lightgrey" stroke="black" points="70,-52 16,-52 16,-16 70,-16 70,-52"></polygon>
<text text-anchor="middle" x="43" y="-29.8" font-family="Times,serif" font-size="14.00">tata</text>
</g>
<!-- tata2 -->
<g id="node2" class="node">
<title>tata2</title>
<polygon fill="lightgrey" stroke="black" points="163,-52 109,-52 109,-16 163,-16 163,-52"></polygon>
<text text-anchor="middle" x="136" y="-29.8" font-family="Times,serif" font-size="14.00">tata2</text>
</g>
</g>
</svg>
</div>
<p></p></figure><p></p>
</div>
</div>
</div>
<section id="varwrrum-variants-at-regions-where-reference-reads-uniquely-map-status-functional" class="level4" data-number="2.2.6.1">
<h4 data-number="2.2.6.1" class="anchored" data-anchor-id="varwrrum-variants-at-regions-where-reference-reads-uniquely-map-status-functional"><span class="header-section-number">2.2.6.1</span> VARWRRUM: Variants At Regions Where Reference Reads Uniquely Map (STATUS: FUNCTIONAL)</h4>
<p><strong>Variant calling - halpoid reference, diploid genome.</strong></p>
<p>This track prepares the reference (an HaploSSembly / haploid genome assembly) and samples for variant calling. Then it does statistical variant calling (using FreeBayes software), does variant filtering and variant normalisation of called variants; this for each sample individually but also for all samples considered as a single population.<br>
Main output: normalised vcf file</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="./files/VARWRRUM.drawio.svg" class="img-fluid figure-img"></p>
<figcaption>VARWRRUM track overview</figcaption>
</figure>
</div>
<p>The detailed track descriptions is provided <a href="./VARWRRUM_choices">here</a>. In short, the reference is softmasked, at least for low complexity regions, <!-- TODO implement rest ? --> the reads of each sample (diploid) <!-- TODO remove the polyploid option or keep as is ? --> are mapped to it, and variants are called for each sample and globaly as a population, using solely reads that are uniquely mapped, this with <a href="">Freebayes software</a>. Variants are filtered, ensuring that solely variants of good quality are retained, and normalised, which allow to compare position of the call between samples. <!-- TODO read more about normalisation variants, should also maybe add phasing here? --></p>
<p>We implemented different statistics calculation (coverage, presence of mapping tags) at the different stages (pre, during, and after) variant calling and filtering.</p>
<p>As mentioned in <a href="https://github.com/freebayes/freebayes">FREEBAYES repository</a>, &gt; “Users are encouraged to examine their output and both variants which are retained and those they filter out”.</p>
<p>We certainly encourage this practice, as this will allow to adjust filtering criteria, and might also hint about improvement to be gained at the reads mapping filtering and/or assembly masking stages.</p>
<p>References - <a href="https://doi.org/10.5281/zenodo.5127899">Trim Galore citat</a> - <a href="https://arxiv.org/abs/1303.3997">BWA-MEM article</a> - <a href="https://academic.oup.com/gigascience/article/10/2/giab008/6137722?login=false">SAMTOOLS article</a> - <a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0163962">SEQKIT article</a> - <a href="https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0185056">BBTools/BBmask article</a> - <a href="https://academic.oup.com/bioinformatics/article/31/13/2202/196142">VT article</a> - <a href="https://arxiv.org/abs/1207.3907">FREEBAYES article</a> - <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-018-2485-7">purge haplotigs article</a> - <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3198575/">BCFtools article</a></p>
<p>Repositories - <a href="https://github.com/FelixKrueger/TrimGalore">Trim Galore repository</a> - <a href="https://github.com/lh3/bwa">BWA repository</a> - <a href="https://www.htslib.org/">SAMTOOLS website</a> - <a href="https://github.com/shenwei356/seqkit/releases">SEQKIT repository</a> - <a href="https://jgi.doe.gov/data-and-tools/software-tools/bbtools/">BBTools/BBmask website</a> - <a href="https://github.com/atks/vt">VT repository</a> - <a href="https://github.com/freebayes/freebayes">FREEBAYES repository</a> - <a href="https://bitbucket.org/mroachawri/purge_haplotigs/src/master/">purge haplotigs repository</a> - <a href="https://github.com/samtools/bcftools">BCFTools repository</a></p>
<p>Manuals and other resources - <a href="https://www.htslib.org/doc/#manual-pages">SAMTOOLS manual</a> - <a href="https://samtools.github.io/bcftools/bcftools.html">BCFTools manual</a> - <a href="https://genome.sph.umich.edu/wiki/Vt">VT documentation</a></p>
<hr>
</section>
</section>
</section>
<section id="fungen" class="level2" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="fungen"><span class="header-section-number">2.3</span> #### FunGen</h2>
<blockquote class="blockquote">
<p>Origin of the track (ref)</p>
</blockquote>
<div class="cell" data-layout-align="default">
<div class="cell-output-display">
<div>
<p></p><figure class="figure"><p></p>
<div>
<svg width="672" height="480" viewbox="0.00 0.00 69.11 32.80" xmlns="http://www.w3.org/2000/svg" xlink="http://www.w3.org/1999/xlink" style="; max-width: none; max-height: none">
<g id="graph0" class="graph" transform="scale(1 1) rotate(0) translate(4 28.8)">
<title>Pipeline_Overview</title>
<polygon fill="white" stroke="transparent" points="-4,4 -4,-28.8 65.11,-28.8 65.11,4 -4,4"></polygon>
<text text-anchor="middle" x="30.56" y="-8.2" font-family="Times,serif" font-size="14.00">FunGen</text>
</g>
</svg>
</div>
<p></p></figure><p></p>
</div>
</div>
</div>
<!--
TODO

See if can find a way to annotate - even if too different - for diploid / by similarity or
I need to explore what could be done for Sapro

### 7. ANNOTATION
- haploSSembly
- Variants then ... annotation effect ?

-->
<hr>
<!--
TODO URGENT

Determinant pathogenecity
- repeats
  - copy nb variation ?
  - pattern / type classification (homology?)

- snps
  - gwas like with unitigs ? phasing possible hum ...


-->
<hr>
<!--
VISUALISATION

- jbrowse - manage to open files

TRY (need X11)
- https://github.com/kcleal/gw
-->
</section>
<section id="references" class="level2" data-number="2.4">
<h2 data-number="2.4" class="anchored" data-anchor-id="references"><span class="header-section-number">2.4</span> References</h2>
<!-- reference so far 2024-07-30  -->
<p>Bushnell, Brian, Jonathan Rood, and Esther Singer. “BBMerge – Accurate Paired Shotgun Read Merging via Overlap.” PLOS ONE 12, no. 10 (October 26, 2017): e0185056. https://doi.org/10.1371/journal.pone.0185056.</p>
<p>Camacho, Christiam, George Coulouris, Vahram Avagyan, Ning Ma, Jason Papadopoulos, Kevin Bealer, and Thomas L. Madden. “BLAST+: Architecture and Applications.” BMC Bioinformatics 10 (December 15, 2009): 421. https://doi.org/10.1186/1471-2105-10-421.</p>
<p>Danecek, Petr, James K Bonfield, Jennifer Liddle, John Marshall, Valeriu Ohan, Martin O Pollard, Andrew Whitwham, et al.&nbsp;“Twelve Years of SAMtools and BCFtools.” GigaScience 10, no. 2 (February 1, 2021): giab008. https://doi.org/10.1093/gigascience/giab008.</p>
<p>Ewels, Philip, Måns Magnusson, Sverker Lundin, and Max Käller. “MultiQC: Summarize Analysis Results for Multiple Tools and Samples in a Single Report.” Bioinformatics 32, no. 19 (October 1, 2016): 3047–48. https://doi.org/10.1093/bioinformatics/btw354.</p>
<p>García-Alcalde, Fernando, Konstantin Okonechnikov, José Carbonell, Luis M. Cruz, Stefan Götz, Sonia Tarazona, Joaquín Dopazo, Thomas F. Meyer, and Ana Conesa. “Qualimap: Evaluating next-Generation Sequencing Alignment Data.” Bioinformatics 28, no. 20 (October 15, 2012): 2678–79. https://doi.org/10.1093/bioinformatics/bts503.</p>
<p>Gurevich, Alexey, Vladislav Saveliev, Nikolay Vyahhi, and Glenn Tesler. “QUAST: Quality Assessment Tool for Genome Assemblies.” Bioinformatics 29, no. 8 (April 15, 2013): 1072–75. https://doi.org/10.1093/bioinformatics/btt086.</p>
<p>Krueger, Felix, Frankie James, Phil Ewels, Ebrahim Afyounian, and Benjamin Schuster-Boeckler. “FelixKrueger/TrimGalore: V0.6.7 - DOI via Zenodo.” Zenodo, July 23, 2021. https://doi.org/10.5281/zenodo.5127899.</p>
<p>Li, Heng. “A Statistical Framework for SNP Calling, Mutation Discovery, Association Mapping and Population Genetical Parameter Estimation from Sequencing Data.” Bioinformatics 27, no. 21 (November 1, 2011): 2987–93. https://doi.org/10.1093/bioinformatics/btr509.</p>
<p>“Aligning Sequence Reads, Clone Sequences and Assembly Contigs with BWA-MEM.” arXiv, May 26, 2013. https://doi.org/10.48550/arXiv.1303.3997.</p>
<p>“Minimap2: Pairwise Alignment for Nucleotide Sequences.” Bioinformatics 34, no. 18 (September 15, 2018): 3094–3100. https://doi.org/10.1093/bioinformatics/bty191.</p>
<p>Quarto. “Quarto.” Accessed July 30, 2024. https://quarto.org/.</p>
<p>“R Core Team (2024). R: A Language and Environment for Statistical Computing. R Foundation for Statistical Computing, Vienna, Austria. URL Https://Www.R-Project.Org/,” n.d.</p>
<p>Roach, Michael J., Simon A. Schmidt, and Anthony R. Borneman. “Purge Haplotigs: Allelic Contig Reassignment for Third-Gen Diploid Genome Assemblies.” BMC Bioinformatics 19, no. 1 (November 29, 2018): 460. https://doi.org/10.1186/s12859-018-2485-7.</p>
<p>Shen, Wei, Shuai Le, Yan Li, and Fuquan Hu. “SeqKit: A Cross-Platform and Ultrafast Toolkit for FASTA/Q File Manipulation.” PLOS ONE 11, no. 10 (October 5, 2016): e0163962. https://doi.org/10.1371/journal.pone.0163962.</p>
<p>“SeqKit: A Cross-Platform and Ultrafast Toolkit for FASTA/Q File Manipulation.” PLOS ONE 11, no. 10 (October 5, 2016): e0163962. https://doi.org/10.1371/journal.pone.0163962.</p>
<p>Simão, Felipe A., Robert M. Waterhouse, Panagiotis Ioannidis, Evgenia V. Kriventseva, and Evgeny M. Zdobnov. “BUSCO: Assessing Genome Assembly and Annotation Completeness with Single-Copy Orthologs.” Bioinformatics 31, no. 19 (October 1, 2015): 3210–12. https://doi.org/10.1093/bioinformatics/btv351.</p>
<p>Tan, Adrian, Gonçalo R. Abecasis, and Hyun Min Kang. “Unified Representation of Genetic Variants.” Bioinformatics 31, no. 13 (July 1, 2015): 2202–4. https://doi.org/10.1093/bioinformatics/btv112.</p>
<p>Wood, Derrick E., Jennifer Lu, and Ben Langmead. “Improved Metagenomic Analysis with Kraken 2.” Genome Biology 20, no. 1 (November 28, 2019): 257. https://doi.org/10.1186/s13059-019-1891-0.</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./requirements_installation.html" class="pagination-link" aria-label="Requirements and installation">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Requirements and installation</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./usage.html" class="pagination-link" aria-label="Input and usage">
        <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Input and usage</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>